File: Java/src/main/java/br/com/locaweb/locamail/api/repository/EmailRepository.java
package br.com.locaweb.locamail.api.repository;

import br.com.locaweb.locamail.api.model.Email;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface EmailRepository extends JpaRepository<Email, Long> {

    @Query(value = "SELECT DISTINCT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE editavel = 0 AND arquivado = 0 AND excluido = 0 AND spam = 0", nativeQuery = true)
    List<Object[]> listarTodosEmails();

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND remetente = :remetente AND editavel = 0 AND arquivado = 0 AND excluido = 0 AND spam = 0", nativeQuery = true)
    List<Object[]> listarEmailsEnviadosPorRemetente(@Param("remetente") String remetente, @Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND editavel = 0 AND arquivado = 0 AND excluido = 0 AND spam = 0 AND id_pasta IS NULL", nativeQuery = true)
    List<Object[]> listarEmailsPorDestinatario(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND editavel = 0", nativeQuery = true)
    List<Object[]> listarEmailsAi(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL WHERE remetente = :remetente AND editavel = 1", nativeQuery = true)
    List<Email> listarEmailsEditaveisPorRemetente(@Param("remetente") String remetente);

    @Query(value = "SELECT * FROM T_LCW_EMAIL WHERE id_email = :id_email", nativeQuery = true)
    Email listarEmailPorId(@Param("id_email") Long id_email);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND importante = 1 AND editavel = 0 AND arquivado = 0 AND excluido = 0 AND spam = 0", nativeQuery = true)
    List<Object[]> listarEmailsImportantesPorIdUsuario(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND agenda_atrelada = 1 AND editavel = 0 AND arquivado = 0 AND excluido = 0 AND spam = 0", nativeQuery = true)
    List<Object[]> listarEmailsSociaisPorIdUsuario(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND arquivado = 1 AND excluido = 0 AND spam = 0 AND editavel = 0", nativeQuery = true)
    List<Object[]> listarEmailsArquivadosPorIdUsuario(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND spam = 1 AND editavel = 0 AND excluido = 0", nativeQuery = true)
    List<Object[]> listarEmailsSpamPorIdUsuario(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND excluido = 1 AND editavel = 0", nativeQuery = true)
    List<Object[]> listarEmailsLixeiraPorIdUsuario(@Param("id_usuario") Long id_usuario);

    @Query(value = "SELECT * FROM T_LCW_EMAIL INNER JOIN T_LCW_ALTERACAO ON T_LCW_EMAIL.id_email = T_LCW_ALTERACAO.alt_id_email WHERE T_LCW_ALTERACAO.alt_id_usuario = :id_usuario AND id_pasta = :id_pasta AND arquivado = 0 AND excluido = 0 AND spam = 0 AND editavel = 0", nativeQuery = true)
    List<Object[]> listarEmailsPorPastaEIdUsuario(@Param("id_usuario") Long id_usuario, @Param("id_pasta") Long id_pasta);
}

File: Kotlin/app/src/main/java/br/com/fiap/locawebmailapp/database/repository/EmailRepository.kt
package br.com.fiap.locawebmailapp.database.repository

import android.content.Context
import android.util.Log
import br.com.fiap.locawebmailapp.database.dao.InstanceDatabase
import br.com.fiap.locawebmailapp.model.Email
import br.com.fiap.locawebmailapp.model.EmailComAlteracao
import br.com.fiap.locawebmailapp.utils.stringParaLista

class EmailRepository(context: Context) {
    private val emailDao = InstanceDatabase.getDatabase(context).emailDao()

    fun criarEmail(email: Email): Long {
        return emailDao.criarEmail(email)
    }

    fun excluirEmail(email: Email) {
        return emailDao.excluirEmail(email)
    }

    fun atualizarEmail(email: Email) {
        return emailDao.atualizarEmail(email)
    }

    fun listarTodosEmails(usuarioRepository: UsuarioRepository): List<EmailComAlteracao> {
        val listTodosEmails = arrayListOf<EmailComAlteracao>()

        for (email in emailDao.listarTodosEmails()) {

            val usuario = usuarioRepository.retornaUsarioPorEmail(email.remetente)

            val idRemetente = if (usuario != null) usuario.id_usuario else null

            if (listTodosEmails.isNotEmpty() && idRemetente != null) {
                if (email.alt_id_usuario != idRemetente && listTodosEmails.last().alt_id_email != email.alt_id_email) {
                    listTodosEmails.add(email)
                }
            } else if (idRemetente != null) {
                if (email.alt_id_usuario != idRemetente) {
                    listTodosEmails.add(email)
                }
            }
        }
        return listTodosEmails
    }

    fun listarEmailsEnviadosPorRemetente(
        remetente: String,
        id_usuario: Long
    ): List<EmailComAlteracao> {
        return emailDao.listarEmailsEnviadosPorRemetente(remetente, id_usuario)
    }


    fun listarEmailsPorDestinatario(
        destinatario: String,
        id_usuario: Long,
        respostaEmailRepository: RespostaEmailRepository
    ): List<EmailComAlteracao> {
        val emailListDb = emailDao.listarEmailsPorDestinatario(id_usuario)

        return emailListDb.filter { email ->
            val respostaEmailList =
                respostaEmailRepository.listarRespostasEmailPorIdEmail(email.id_email)

            stringParaLista(email.destinatario).contains(destinatario) ||
                    stringParaLista(email.cc).contains(destinatario) ||
                    stringParaLista(email.cco).contains(destinatario) ||
                    respostaEmailList.any {respostaEmail ->
                        stringParaLista(respostaEmail.destinatario).contains(destinatario) ||
                                stringParaLista(respostaEmail.cc).contains(destinatario) ||
                                stringParaLista(respostaEmail.cco).contains(destinatario)
                    }
        }
    }

    fun listarEmailPorId(id_email: Long): Email {
        return emailDao.listarEmailPorId(id_email)
    }

    fun listarEmailsEditaveisPorRemetente(remetente: String): List<Email> {
        return emailDao.listarEmailsEditaveisPorRemetente(remetente)
    }


    fun listarEmailsPorPastaEIdUsuario(id_usuario: Long, id_pasta: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsPorPastaEIdUsuario(id_usuario, id_pasta)
    }

    fun listarEmailsImportantesPorIdUsuario(id_usuario: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsImportantesPorIdUsuario(id_usuario)
    }

    fun listarEmailsAi(id_usuario: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsAi(id_usuario)
    }

    fun listarEmailsArquivadosPorIdUsuario(id_usuario: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsArquivadosPorIdUsuario(id_usuario)
    }

    fun listarEmailsLixeiraPorIdUsuario(id_usuario: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsLixeiraPorIdUsuario(id_usuario)
    }

    fun listarEmailsSpamPorIdUsuario(id_usuario: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsSpamPorIdUsuario(id_usuario)
    }

    fun listarEmailsSociaisPorIdUsuario(id_usuario: Long): List<EmailComAlteracao> {
        return emailDao.listarEmailsSociaisPorIdUsuario(id_usuario)
    }

}